### PR TITLE
fix: avoid crash by converting lateinit to nullable

### PR DESCRIPTION
**Pull request by mikehardy**
## Purpose / Description

Avoid a crash seen in ACRA by the evil use of `lateinit` which is just the sleeper NullPointerException of modern Kotlin times

## Fixes

https://ankidroid.org/acra/app/1/bug/260934/report/c4097fac-9c67-460b-986c-017cc1a98313

```
com.ichi2.anki.exception.UserSubmittedException: lateinit property dropDownDecks has not been initialized
	at com.ichi2.anki.DeckSpinnerSelection.selectDeck(DeckSpinnerSelection.kt:201)
	at com.ichi2.anki.DeckSpinnerSelection.selectDeckById(DeckSpinnerSelection.kt:190)
	at com.ichi2.anki.CardBrowser.setupFlows$onDeckIdChanged(CardBrowser.kt:465)
	at com.ichi2.anki.CardBrowser.access$setupFlows$onDeckIdChanged(CardBrowser.kt:155)
	at com.ichi2.anki.CardBrowser$setupFlows$7.invoke(CardBrowser.kt:570)
	at com.ichi2.anki.CardBrowser$setupFlows$7.invoke(CardBrowser.kt:570)
	at com.ichi2.anki.CardBrowser$launchCollectionInLifecycleScope$1$1.emit(CardBrowser.kt:2509)
	at kotlinx.coroutines.flow.StateFlowImpl.collect(StateFlow.kt:396)
	at com.ichi2.anki.CardBrowser$launchCollectionInLifecycleScope$1.invokeSuspend(CardBrowser.kt:2502)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.internal.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:363)
	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:26)
	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable$default(Cancellable.kt:21)
	at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:88)
	at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:123)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:52)
	at kotlinx.coroutines.BuildersKt.launch(Builders.kt:1)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:43)
	at kotlinx.coroutines.BuildersKt.launch$default(Builders.kt:1)
	at com.ichi2.anki.CardBrowser.launchCollectionInLifecycleScope(CardBrowser.kt:2501)
	at com.ichi2.anki.CardBrowser.setupFlows(CardBrowser.kt:570)
	at com.ichi2.anki.CardBrowser.onCreate(CardBrowser.kt:426)
	at android.app.Activity.performCreate(Activity.java:8207)
	at android.app.Activity.performCreate(Activity.java:8191)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1309)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3811)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:4020)
	at android.app.ActivityThread.handleRelaunchActivityInner(ActivityThread.java:6056)
	at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:5951)
	at android.app.servertransaction.ActivityRelaunchItem.execute(ActivityRelaunchItem.java:69)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
	at android.app.ClientTransactionHandler.executeTransaction(ClientTransactionHandler.java:63)
	at android.app.ActivityThread.handleRelaunchActivityLocally(ActivityThread.java:6015)
	at android.app.ActivityThread.access$3500(ActivityThread.java:301)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2338)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loop(Looper.java:246)
	at android.app.ActivityThread.main(ActivityThread.java:8633)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [p0{Cancelling}@afaf4a2, Dispatchers.Main.immediate]

```

* Fixes #

## Approach

There is apparently an init pathway where the flows for CardBrowser start up a DeckSpinner and that can try to select where it should be, *before* `onCollectionLoaded` is finished, so the spinner doesn't have all the data it expects because it hasn't actually been initialized fully.

Instead of just blindly expecting itself to be initialized now, the deck list in the spinner is nullable, and if it is null, then the spinner just defaults to all decks / position -1

I assume when it is initialized that the spinner will set itself up correctly later, but most importantly it won't crash now

The diff shouldn't look so bad if ignore space is turned on, because mostly I'm encapsulating the init blocks in `let {}` to handle the new nullability cleanly

## How Has This Been Tested?

open the app with this patch installed, go to card browser and check the spinner is still working, it's still working

## Learning (optional, can help others)

Anyone accepting PRs that have lateinit, or `!!` etc are asking for trouble 🧐 

## Checklist
_Please, go through these checks before submitting the PR._

- [ ] You have a descriptive commit message with a short title (first line, max 50 chars).
- [ ] You have commented your code, particularly in hard-to-understand areas
- [ ] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by BrayanDSO — APPROVED** [review]
The code doesn't look pretty, but not crashing is better than crashing

**Comment by mikehardy** [comment]
Yeah, I wasn't sure how to handle it without doing a lot of either `if (foo !== null)` or similar, but I really just wanted to make it not lateinit so that static typing would ensure I got every single spot it might be racy-not-initialized. Was the best I came up with, using my baby kotlin skills

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckSpinnerSelection.kt [hunk 1]
```diff
@@ -184,7 +195,7 @@ class DeckSpinnerSelection(
      * @return True if selection succeeded.
      */
     suspend fun selectDeckById(deckId: DeckId, setAsCurrentDeck: Boolean): Boolean {
-        return if (deckId == ALL_DECKS_ID) {
+        return if (deckId == ALL_DECKS_ID || this.dropDownDecks != null) {
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckSpinnerSelection.kt:198]: Shouldn't the condition be _this.dropDownDecks == null_? Right now it reads that if the deck id is ALL_DECKS_ID and the list of decks is available(which will be like 99,9% of the situations) then select all decks.

**Review by lukstbit — CHANGES_REQUESTED** [review]
With the changes in this PR when I go in the card browser and select different decks all I see in the toolbar is the title _All Decks_(it never changes) and the number of cards of the deck that I actually selected in the dialog.

The whole code around deck selection really needs to be refactored, it's just a mess in its current form.

**Comment by BrayanDSO** [comment]
> The whole code around deck selection really needs to be refactored

agreed

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckSpinnerSelection.kt (hunk 1 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckSpinnerSelection.kt]: Oh wow, that's a small but important mistake. Thanks @lukstbit
This was the root of the repro steps you gave showing all decks every time, I just pushed the `==` vs `!=` version of this which is obviously the right conditional, and now every where I try to use the deck spinner (add note, card browser, statistics) it appears to work correctly.
Sorry about that

**Review by criticalAY — APPROVED** [review]
LGTM!

**Review by BrayanDSO — APPROVED** [review]

**Review by lukstbit — APPROVED** [review]
Everything seems to be in order now!

**Pull request merged**